src/node/tendermint_internals.ml

```diff
@@ -0,0 +1,158 @@
+  | Block b -> Block (Protocol.Block.update_round b ~consensus_round:round)
+
+type height = int64 [@@deriving yojson]
+
+type round = int [@@deriving yojson]
+
+type consensus_step =
+  | Proposal
+  | Prevote
+  | Precommit
+
+let string_of_step = function
+  | Proposal -> "Proposal"
+  | Prevote -> "Prevote"
+  | Precommit -> "Precommit"
+
+type sidechain_consensus_op =
+  | ProposalOP  of (height * round * value * round)
+  | PrevoteOP   of (height * round * value)
+  | PrecommitOP of (height * round * value)
+[@@deriving yojson]
+
+let step_of_op = function
+  | ProposalOP _ -> Proposal
+  | PrevoteOP _ -> Prevote
+  | PrecommitOP _ -> Precommit
+
+let string_of_op = function
+  | ProposalOP (height, round, value, vround) ->
+    Printf.sprintf "<PROPOSAL, %Ld, %d, %s, %d>" height round
+      (string_of_value value) vround
+  | PrevoteOP (height, round, value) ->
+    Printf.sprintf "<PREVOTE, %Ld, %d, %s>" height round (string_of_value value)
+  | PrecommitOP (height, round, value) ->
+    Printf.sprintf "<PRECOMMIT, %Ld, %d, %s>" height round
+      (string_of_value value)
+
+let height = function
+  | ProposalOP (h, _, _, _)
+  | PrevoteOP (h, _, _)
+  | PrecommitOP (h, _, _) ->
+    h
+
+let round = function
+  | ProposalOP (_, r, _, _)
+  | PrevoteOP (_, r, _)
+  | PrecommitOP (_, r, _) ->
+    r
+
+let value_of_op = function
+  | ProposalOP (_, _, v, _)
+  | PrevoteOP (_, _, v)
+  | PrecommitOP (_, _, v) ->
+    v
+
+let hash_of_value = function
+  | Nil -> Crypto.BLAKE2B.hash ""
+  | Block b -> b.Block.hash
+
+let hash_of_consensus_op consensus_op sender =
+  let s1 = string_of_op consensus_op in
+  let s2 = Crypto.Key_hash.to_string sender in
+  Crypto.BLAKE2B.hash (s1 ^ s2)
+
+let hash_of_consensus_value consensus_op =
+  let _, _, value =
+    match consensus_op with
+    | ProposalOP (h, r, v, _)
+    | PrevoteOP (h, r, v)
+    | PrecommitOP (h, r, v) ->
+      (h, r, v) in
+  hash_of_value value
+
+type consensus_state = {
+  mutable height : height;
+  mutable round : round;
+  mutable step : consensus_step;
+  mutable locked_value : value;
+  mutable locked_round : round;
+  mutable valid_value : value;
+  mutable valid_round : round;
+}
+
+let fresh_state height =
+  {
+    height;
+    round = 0;
+    step = Proposal;
+    locked_value = nil;
+    locked_round = -1;
+    valid_value = nil;
+    valid_round = -1;
+  }
+
+let short name =
+  let name = Crypto.Key_hash.to_string name in
+  String.sub name (String.length name - 6) 6
+
+let debug state msg =
+  let self = Crypto.Key_hash.to_string state.State.identity.t in
+  let self = String.sub self (String.length self - 6) 6 in
+  prerr_endline ("*** " ^ self ^ "   " ^ msg)
+
+let is_allowed_proposer (global_state : State.t) (height : height)
+    (round : round) (address : Key_hash.t) =
+  let protocol_state = global_state.protocol in
+  let proposer = proposer protocol_state.validators height round in
+  let b = Key_hash.equal address proposer.address in
```

Comment:
I don't believe `let b = <bool expr> in b` is any more meaningful than simply `<bool expr>` in this case.